### ISSUE TITLE
`edge.controlPoints()` is `undefined` for loops

### ISSUE THREAD

**Issue by josejulio** [cytoscape/cytoscape.js#2139]
Bug report

tl;dr; `edgeLoop.controlPoints()` returns falsy for loop edges.

**Environment info**

- Cytoscape.js version :3.2.12
- Browser/Node.js & version : Chrome 67.0.3396.79 (Official Build) (64-bit)

**Current (buggy) behaviour**

edge loops don't have their control points exposed, but they are in `_private.rscratch.ctrlpts`

**Desired behaviour**

edge loops should return their control points used.

**Minimum steps to reproduce**

Create a graph with a edge loop, try to get its controlPoints

http://jsbin.com/nenawutewi/1/edit?html,css,js,console,output

So maybe i'm missunderstanding something, but if not, i think [this](https://github.com/cytoscape/cytoscape.js/blob/master/src/extensions/renderer/base/coord-ele-math/edge-control-points.js#L789) should also check for 'self' type

**Info on what we are trying to do**

We're using `cytoscapejs` in [Kiali](https://kiali.org/) to create a graph for a service mesh, currently we are trying to generate a way to graphically represent traffic, so i'm using the [cytoscape-canvas](https://github.com/classcraft/cytoscape.js-canvas) extension to get a new canvas and draw shapes to represent said traffic.

I'm using the source, control points and traffic to compute the bezier path and draw the balls on the edge path.

On node-to-node edges, this looks good but have found that loop/self edges don't have exposed the controlPoints. Looking at the code i have seen that they do indeed have controlPoints in `theEdge._private.rscratch.ctrlpts`

edit: I noticed that the ctrlpoints are for a quadratic beziers and not cubics, using start, controlpoint_1 and midpoint draw correctly the midpath.
~~Using these points don't follow the "right" path (its a bit off)~~

~~(red diamonds are controlPoints and green is the "midpoint" (not used to compute the bezier)~~
![loop-cordpoints](https://user-images.githubusercontent.com/3845764/41564924-29713cac-731a-11e8-9d24-6e0accc01b28.gif)

~~I also tried using `theEdge._private.rscratch.allpts` but with no luck (although it looks closer):~~
![loop-allpts](https://user-images.githubusercontent.com/3845764/41565650-3ddc2e60-731d-11e8-8f5c-8127ce508513.gif)

~~I looked at [this code](https://github.com/cytoscape/cytoscape.js/blob/master/src/extensions/renderer/canvas/drawing-edges.js#L148) and it draws `bezier` and `self` alike, so maybe i'm missing something?~~

**Comment by maxkfranz** [comment]
All curved edges are a series of one or more quadratic beziers.  The t=1 point for the nth bezier is used as the t=0 point for the (n+1)th bezier.  So for loops, the first bezier has t=0 at the node and t=1 at the midpoint.  The second bezier has t=0 at the midpoint and t=1 at the node.

**Comment by josejulio** [comment]
> All curved edges are a series of one or more quadratic beziers. The t=1 point for the nth bezier is used as the t=0 point for the (n+1)th bezier. So for loops, the first bezier has t=0 at the node and t=1 at the midpoint. The second bezier has t=0 at the midpoint and t=1 at the node.

Thank for the explanation!

**Comment by josejulio** [comment]
Question, could we also expose the "allPoints" ? Or do i need to compute the all points (as done [here](https://github.com/cytoscape/cytoscape.js/blob/master/src/extensions/renderer/base/coord-ele-math/edge-control-points.js#L638-L650)) 
Thanks

**Comment by josejulio** [comment]
Thanks, I'll currently stick to computing the values from the controlPoints.
Is there edges that have more than 3 control points (such as the loop)? If so, does that depend on the layout or to decide or is an internal of cytoscapejs?

**Comment by maxkfranz** [comment]
Only loops have more than one by default.  For non-loop edges, it's only more than one if you specify more than one control point.
